pkg/asset/store.go

```diff
@@ -14,47 +20,212 @@ type Store interface {
+	onDiskFiles map[string][]byte
+	assets      map[reflect.Type]Asset
+	// dirtyAssets marks whether or not the asset is dirty.
+	dirtyAssets map[reflect.Type]bool
+}
+
+// NewStore returns an asset store that implements the Store interface.
+func NewStore(dir string) (Store, error) {
+	fileMap, err := readAllFiles(dir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read files from dir %q: %v", dir, err)
+	}
+	return &StoreImpl{
+		onDiskFiles: fileMap,
+		assets:      make(map[reflect.Type]Asset),
+		dirtyAssets: make(map[reflect.Type]bool),
+	}, nil
 }
 
 // Fetch retrieves the state of the given asset, generating it and its
 // dependencies if necessary.
 func (s *StoreImpl) Fetch(asset Asset) error {
-	return s.fetch(asset, "")
+	_, err := s.fetch(asset, "")
+	return err
 }
 
-func (s *StoreImpl) fetch(asset Asset, indent string) error {
+// fetch retrieves the state of the given asset, generating it and its
+// dependencies if necessary.
+// It returns dirty if the asset or any of its parents is loaded from on-disk files.
+func (s *StoreImpl) fetch(asset Asset, indent string) (dirty bool, err error) {
 	logrus.Debugf("%sFetching %s...", indent, asset.Name())
+
+	// Return immediately if the asset is found in the cache,
+	// this is because we are doing a depth-first-search, it's guaranteed
+	// that we always fetch the parent before children, so we don't need
+	// to worry about invalidating anything in the cache.
 	storedAsset, ok := s.assets[reflect.TypeOf(asset)]
 	if ok {
 		logrus.Debugf("%sFound %s...", indent, asset.Name())
 		reflect.ValueOf(asset).Elem().Set(reflect.ValueOf(storedAsset).Elem())
-		return nil
+		return s.dirtyAssets[reflect.TypeOf(asset)], nil
 	}
 
 	dependencies := asset.Dependencies()
 	parents := make(Parents, len(dependencies))
 	if len(dependencies) > 0 {
 		logrus.Debugf("%sGenerating dependencies of %s...", indent, asset.Name())
 	}
+
+	var anyDirty bool
 	for _, d := range dependencies {
-		err := s.fetch(d, indent+"  ")
+		dt, err := s.fetch(d, indent+"  ")
 		if err != nil {
-			return errors.Wrapf(err, "failed to fetch dependency for %s", asset.Name())
+			return false, errors.Wrapf(err, "failed to fetch dependency for %s", asset.Name())
+		}
+		if dt {
+			anyDirty = true
 		}
 		parents.Add(d)
 	}
 
-	logrus.Debugf("%sGenerating %s...", indent, asset.Name())
-	err := asset.Generate(parents)
-	if err != nil {
-		return errors.Wrapf(err, "failed to generate asset %s", asset.Name())
+	var found bool
+	loadable, ok := asset.(LoadableAsset)
+	if ok {
+		found, err = loadable.Load(LoadFileWith(s.onDiskFiles))
+		if err != nil {
+			return false, errors.Wrapf(err, "unexpected error when loading asset %s", asset.Name())
+		}
 	}
-	if s.assets == nil {
-		s.assets = make(map[reflect.Type]Asset)
+
+	// Try to load from on-disk files first.
+	switch {
+	case anyDirty && found:
+		return false, fmt.Errorf("cannot resolve, both parent assets and current asset %q are dirty", asset.Name())
```

Comment:
Do we have a plan for cleaning up files left by previous steps? For example, if the user runs `openshift-install install-config` and `openshift-install manifests` before running `openshift-install cluster`, then the last one will fail. The manifests need to be re-generated because install-config.yml exists, but the manifest files exist on disk as well.

Reply:
I think the plan is to remove the install-config.yml after running openshift-install manifests.